### PR TITLE
Fix a bug about duplicate validator key

### PR DESCRIPTION
**Pull request by gguoss**
Thank you for your Pull Request!

Before you submitting, please check that:

- [ ] You added a brief description of the PR, e.g.:
  - What does it do?
  - What important points reviewers should know?
  - Is there something left for follow-up PRs?
- [ ] You labeled the PR with appropriate labels if you have permissions to do so.
- [ ] You mentioned a related issue if this PR related to it, e.g. `Fixes #228` or `Related #1337`.
- [ ] You asked any particular reviewers to review. If you aren't sure, start with GH suggestions.
- [ ] Your PR adheres [the style guide](https://github.com/paritytech/polkadot/wiki/Style-Guide)
  - In particular, mind the maximal line length.
  - There is no commented code checked in unless necessary.
  - Any panickers have a proof or removed.
- [x] You bumped the runtime version if there are breaking changes in the **runtime**.
- [ ] You updated any rustdocs which may have changed

After you've read this notice feel free to remove it.
Thank you!

✄ -----------------------------------------------------------------------------

### CODE REVIEW

File: srml/session/src/lib.rs [hunk 1]
```diff
@@ -165,8 +165,11 @@ decl_module! {
 		/// This doesn't take effect until the next session.
 		fn set_key(origin, key: T::SessionKey) {
 			let who = ensure_signed(origin)?;
-			// set new value for next session
-			<NextKeyFor<T>>::insert(who, key);
+			if <KeyFilterMap<T>>::get(&key) == None {
```
Comment (gavofyork) [inline srml/session/src/lib.rs]:
```suggestion
			if <KeyFilterMap<T>>::get(&key).is_none() {
```

**Comment by gavofyork** [comment]
what bug is this meant to fix?

**Comment by gguoss** [comment]
> what bug is this meant to fix?

Validator (A, B ).           
- A:    AuthorityKey (  authA )
- B:  also set AuthorityKey ( authA), then B can't run a node. 

Validator's key will be two same AuthorityKey.

**Comment by gavofyork** [comment]
please explain further.

**Comment by bkchr** [comment]
@gguoss ping

**Comment by gguoss** [comment]
Registering the authority node does not require any physical machine to run the node. Because I just need to set my authority key to the existing authority key of the running node. This way I can get dividends as usual.

**Comment by gguoss** [comment]
> Registering the authority node does not require any physical machine to run the node. Because I just need to set my authority key to the existing authority key of the running node. This way I can get dividends as usual.

The reason for doing evil is: A duplicate authority key can be added to the aura consensus.

File: srml/session/src/lib.rs [hunk 2]
```diff
@@ -165,8 +165,11 @@ decl_module! {
 		/// This doesn't take effect until the next session.
 		fn set_key(origin, key: T::SessionKey) {
 			let who = ensure_signed(origin)?;
```
Comment (gavofyork) [inline srml/session/src/lib.rs]: it should remove the current entry for `NextKeyFor(who)` in `KeyFilterMap`.

**Comment by gavofyork** [comment]
so this is to prevent two validators assigning themselves the same session key?

**Comment by gguoss** [comment]
> so this is to prevent two validators assigning themselves the same session key?

yes

**Comment by gavofyork** [comment]
This is superseded now.

**Pull request closed without merging**